agent/agent.go

```diff
@@ -507,6 +511,63 @@ func (a *Agent) UpdateTaskStatus(ctx context.Context, taskID string, status *api
 
 		return nil
 	}); err != nil {
+		log.G(ctx).Debug("failed to send task status")
+		return err
+	}
+
+	select {
+	case err := <-errs:
+		return err
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// UpdateTaskStatuses performs a bulk update of ALL task statuses provided.
+func (a *Agent) UpdateTaskStatuses(ctx context.Context, statuses map[string]*api.TaskStatus) error {
+	ctx = log.WithField(ctx, "method", "(*Agent).UpdateTaskStatuses")
+	log.G(ctx).Debugf("Updating %v task statuses", len(statuses))
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	errs := make(chan error, 1)
+	if err := a.withSession(ctx, func(session *session) error {
+		go func() {
+			// convert statuses into the actual API object we need
+			updates := make([]*api.UpdateTaskStatusRequest_TaskStatusUpdate, 0, len(statuses))
+			for id, status := range statuses {
+				updates = append(updates, &api.UpdateTaskStatusRequest_TaskStatusUpdate{
+					TaskID: id,
+					Status: status,
+				})
+			}
+
+			var err error
+			for len(updates) > 0 {
+				// session.sendTaskStatuses can only send 1024 updates at a time.
```

Comment:
Sending more than what the dispatcher can process seems suboptimal. Since this is a closed system, it's probably OK to make the agent aware of this limit and split the batch up on the sending side.

Reply:
I'd say "no" only because this is the pattern hinted at by the `session.sendTaskStatuses` method, and lets us ignore how large the batch CAN be in favor of just sending as much as we want and looping.